File: programmers/가장_큰_정사각형_찾기.java
class Solution
{
    public int solution(int [][]board)
    {

		int[][] newBoard = new int[board.length+1][board[0].length+1];
		
		for ( int i = 0 ; i < board.length; i++ ) {
			for ( int j = 0 ; j < board[0].length; j++ ) {
				newBoard[i+1][j+1] = board[i][j];
			}
		}
		
		int max = 0;
		for ( int i = 1 ; i < newBoard.length; i ++ ) {
			for ( int j = 1 ; j < newBoard[i].length; j++ ) {

				if ( newBoard[i][j] >= 1 ) {
					
					int left = newBoard[i-1][j];
					int up = newBoard[i][j-1];
					int leftDiagonal = newBoard[i-1][j-1];

					int min = Math.min(left, up);
					min = Math.min(min, leftDiagonal);
					newBoard[i][j] = min+1;
					max = Math.max(max, min+1);
				}
			}
		}
		return (int)Math.pow(max, 2);
	
    }
}

File: programmers/level2/가장_큰_정사각형_찾기.cpp
#include <vector>
using namespace std;

int solution(vector<vector<int>> board)
{
    int answer = board[0][0]; // 정사각형 변의 길이
    int m = board.size();
    int n = board[0].size(); // m x n 배열
    
    for (int i=1; i<m; ++i) { // boundary
        for (int j=1; j<n; ++j) {
            if (board[i][j] == 1) { // 아직 방문 안 한 칸. 0은 정사각형에 포함될 수 X
                board[i][j] = 1 + min(board[i][j-1], min(board[i-1][j-1], board[i-1][j])); // 왼쪽 칸, 왼쪽 위 칸, 위 칸의 값 비교하는 DP 방식, 기존 정사각형 변의 길이에서 더 큰 정사각형 생기므로 하나 더 늘려줌. 만약 0 포함 시 정사각형 생성 불가하므로 변동 X
                answer = max(answer, board[i][j]); // 최대값 갱신
            }
        }
    }

    return answer * answer;
}